### PR TITLE
Bazel overhauls

### PR DESCRIPTION
**Pull request by dgdonut**
Will go through a pass of refinement, but wanted to get this viewable - assume no substantive changes unless requested.

Addresses #353 #357 #358 #361 

High level changes:
* Added some exec client code, more functionality to bzutil tool that can be used to play with remoting capability (CLI, self-explanatory with --help)
* Added plumbing to support population of more complex GoogleAPI RPC Status data down in the Runner, to be able to provide clients with required "your input dir or command was not uploaded to the CAS" type messages
** Required making bazel thrift data all optional. Do not @ me here, Thrift is stupid and this is the only way to make this work
* Some general tightening of response values to API spec. Ticket is opened to address testing here.
* Hardcoded empty sha support in Filer. I may redo this to avoid having the Filer aware of this, and have hardcoded "blobs" that get handled by the CAS API, since clients could still try to read or write empty sha data directly.
* No scoot headers for bazel tasks

### CODE REVIEW

**Comment by illicitonion** [comment]
Verified that this fixes #353 #357 #358 from the pants side :) Thanks!

File: bazel/cas/service.go [hunk 1]
```diff
@@ -89,19 +95,19 @@ func (s *casServer) BatchUpdateBlobs(
 	return nil, status.Error(codes.Unimplemented, "Currently unsupported in Scoot - update blobs independently")
 }
 
-// DEPRECATED - Included for protobuf generated code compatability/compilation
+// GetTree not supported in Scoot for V1
```
Comment (ryancouto) [inline bazel/cas/service.go:102]: Is this confirmed as being part of the API spec?

File: binaries/bzutil/main.go [hunk 2]
```diff
@@ -5,38 +5,101 @@ package main
...
+
 func main() {
 	log.AddHook(hooks.NewContextHook())
 
-	// TODO include optional directory that gets BzFiler.Ingest'ed, so we have a tool that easily satisfies prereqs
+	// TODO additional commands
+	// - upload Directory (tbd)
+	// - download outputs, stdout/err
+
+	// Subcommands
```
Comment (ryancouto) [inline binaries/bzutil/main.go]: nit: this main() is a little long, consider splitting into addSubcommands, addSubcommandFlags, etc.

**Comment by dgdonut** [comment]
this needs a review or approval @ryancouto @JeanetteBruno @jasonbs10

File: runner/runners/bazel.go [hunk 3]
```diff
@@ -34,7 +38,8 @@ func preProcessBazel(filer snapshot.Filer, cmd *runner.Command) error {
 		return fmt.Errorf("Nil ExecuteRequest data in Command with RunType Bazel")
 	}
 	if err := fetchBazelCommand(bzFiler, cmd); err != nil {
-		return fmt.Errorf("Error retrieving Bazel command data: %v", err)
```
Comment (ryancouto) [inline runner/runners/bazel.go:37]: maybe log.Errorf(Error retrieving Bazel command data: %v")

File: runner/runners/bazel.go [hunk 4]
```diff
@@ -34,7 +38,8 @@ func preProcessBazel(filer snapshot.Filer, cmd *runner.Command) error {
...
+		// NB: Important to return this error as-is
```
Comment (ryancouto) [inline runner/runners/bazel.go]: why?

File: runner/runners/bazel.go [hunk 5]
```diff
@@ -79,20 +85,28 @@ func postProcessBazel(filer snapshot.Filer,
 
 	stdoutDigest, err := writeFileToCAS(bzFiler, stdout.AsFile())
 	if err != nil {
-		return nil, fmt.Errorf("Error ingesting stdout to CAS: %s", err)
+		errstr := fmt.Sprintf("Error ingesting stdout to CAS: %s", err)
+		log.Error(errstr)
+		return nil, fmt.Errorf("%s", errstr)
```
Comment (ryancouto) [inline runner/runners/bazel.go]: nit: for all of these, can just fmt.Errorf(errstr)

File: runner/runners/bazel.go [hunk 6]
```diff
@@ -103,6 +117,7 @@ func postProcessBazel(filer snapshot.Filer,
 			StdoutDigest:      stdoutDigest,
 			StderrDigest:      stderrDigest,
 		},
+		ActionDigest: cmd.ExecuteRequest.ActionDigest,
```
Comment (ryancouto) [inline runner/runners/bazel.go]: Do we need to check if cmd.ExecuteRequest is nil to avoid a nil pointer dereference?

File: testCoverage.sh [hunk 7]
```diff
@@ -2,11 +2,15 @@
 
 set -e
 echo "" > coverage.txt
+echo "" > dirlist.out
 
-for d in $(find . -name \*.go | xargs -n 1 dirname | uniq | grep -v /vendor/ | grep -v /fuse); do
+find . -name \*.go | xargs -n 1 dirname | uniq | grep -v /vendor/ | grep -v /fuse > dirlist.out
+
+for d in $(sort dirlist.out | uniq); do
```
Comment (ryancouto) [inline testCoverage.sh:9]: curious - why this change?

File: runner/runners/invoke.go [hunk 8]
```diff
@@ -93,8 +96,23 @@ func (inv *Invoker) run(cmd *runner.Command, id runner.RunID, abortCh chan struc
...
+			// the failure run status that indicates missing data to client
+			if _, ok := err.(*cas.NotFoundError); ok {
+				log.Info("NotFound error for Command during Bazel preprocess - Setting grpc Status error")
+				errStatus, err := getFailedPreconditionStatus([]*remoteexecution.Digest{
+					cmd.ExecuteRequest.Request.GetAction().GetCommandDigest()})
+				if err != nil {
+					log.Errorf("Error generating Failed Precondition status: %s", err)
+				} else {
+					failedStatus.ActionResult = &bazelapi.ActionResult{GRPCStatus: errStatus}
+					return failedStatus
```
Comment (ryancouto) [inline runner/runners/invoke.go]: don't need this, right?

File: runner/runners/invoke.go [hunk 9]
```diff
@@ -158,8 +176,24 @@ func (inv *Invoker) run(cmd *runner.Command, id runner.RunID, abortCh chan struc
...
+			// Status error indicating that the InputRoot data could not be found.
+			if runType == runner.RunTypeBazel {
+				if _, ok := err.(*bzsnapshot.CheckoutNotExistError); ok {
+					log.Info("Checkout for Bazel command returned CheckoutNotExistError - Setting grpc Status error")
+					errStatus, err := getCheckoutMissingStatus(cmd.SnapshotID)
+					if err != nil {
+						log.Errorf("Error generating Failed Precondition status: %s", err)
+					} else {
+						failedStatus.ActionResult = &bazelapi.ActionResult{GRPCStatus: errStatus}
+						return failedStatus
```
Comment (ryancouto) [inline runner/runners/invoke.go]: ditto

File: bazel/cas/service.go (hunk 1 above)
Comment (dgdonut) [reply bazel/cas/service.go:102]: Yes - had this clarified in the Bazel API Revisions  gdoc

File: binaries/bzutil/main.go (hunk 2 above)
Comment (dgdonut) [reply binaries/bzutil/main.go]: added a few comments and grouped flags w/ subcommands to make a bit clearer

File: runner/runners/bazel.go (hunk 3 above)
Comment (dgdonut) [reply runner/runners/bazel.go:37]: added

File: runner/runners/bazel.go (hunk 5 above)
Comment (dgdonut) [reply runner/runners/bazel.go]: done, thanks

File: runner/runners/bazel.go (hunk 4 above)
Comment (dgdonut) [reply runner/runners/bazel.go]: added comments here and in bazel/cas/service about returning NotFoundError. kind of an open question what the standard for errors should be - bubbling up vs wrapping in a new err, when to log, etc.

File: runner/runners/bazel.go (hunk 6 above)
Comment (dgdonut) [reply runner/runners/bazel.go]: We should never hit that case, the first thing we do in the bazel workflow is check that ExecReq is non nil

File: testCoverage.sh (hunk 7 above)
Comment (dgdonut) [reply testCoverage.sh:9]: We have always been testing packages during the coverage step multiple times because we are not uniq'ing between for loop iterations

File: runner/runners/invoke.go (hunk 8 above)
Comment (dgdonut) [reply runner/runners/invoke.go]: removed both, thanks

File: testCoverage.sh (hunk 7 above)
Comment (ryancouto) [reply testCoverage.sh:9]: gotcha - nice catch

File: runner/runners/bazel.go (hunk 6 above)
Comment (ryancouto) [reply runner/runners/bazel.go]: Gotcha. Think it'd still be worth using the helper method cmd.ExecuteRequest.GetActionDigest() from https://github.com/twitter/scoot/pull/362/files#diff-9776fc4ea2a6e05f737bf7c9ef6106bfR58?

**Review by ryancouto — APPROVED** [review]

File: runner/runners/bazel.go (hunk 6 above)
Comment (dgdonut) [reply runner/runners/bazel.go]: ohh, yes, thanks
